### DISCUSSION TITLE
Support for a Passkeys wallet

### DISCUSSION THREAD

**Discussion by TimDaub** [rainbow-me/rainbowkit#1829] [General]
Hey,

with Kiwi News, we're storing a user's temporary upvoting and link submission key in localstorage. But now we're planning to back it up e.g. in iCloud using Passkey's largeBlob.
So e.g. there is now a process to get your key stored in iCloud, in our settings, and technically in the future, it should then be possible to click 'Connect Wallet', but then directly select one of the public keys from iCloud and read the largeBlob to re-instantiate the private key.

So it'd be nice if I could provide an extra wallet option where I can define for myself what shall happen when the user clicks that option. In my case, I'd want to call the navigator.credential API, get the largeBlob, store it in localstorage and then close the connect wallet modal successfully again.

**Comment by 0xmkh** [comment]
@TimDaub We currently don't support passkey wallet. Apologies if this doesn't solve your issue, but let me know if you've got any other question 🙏

**Comment by TimDaub** [comment]
Is there any way I can add a custom wallet button that does an action onClick? I‘ve looked into this, but the wallet connector modules are designed too descriptively, I would have preferred if they were react components, one could insert as children and define functionality. There also doesn‘t seem any other possibility to manipulate the open modal but to define a disclaimer. I‘m using 1.x

**Reply by 0xmkh** [reply]
I see. You can't directly add a react component into the connect modal. What you can do is create your own connector and then add it to RainbowKit. Here is an example of wagmi's [coinbase wallet connector](https://github.com/wevm/wagmi/blob/main/packages/connectors/src/coinbaseWallet.ts)

**Reply by TimDaub** [reply]
That‘s nice and certainly something I‘m interested in doing in the future.

However, at this point for us this is creating a minimal viable prototype using the connect dialogue.

The bummer is that the intent of restoring a temporary key from a passkey storage any connecting your wallet are both behind „Connect Wallet.“ But if e.g. I now click on „Connect Wallet,“ but then at first the modal doesn‘t open but the user is asked to restore the key from Passkeys, I think this can be disruptive to most users‘ interaction with a connect button.

The connector is interesting but all I need at this moment is to add a button that can do a request to credentials.get(), and store the key in localstorage. Any way I can achieve that with the existing code? I don‘t mind breaking framework conventions etc

**Reply by 0xmkh** [reply]
Got it. Yeah that's a bit harder to do. Meanwhile can you look at a way of doing it during the wallet connection ? When you click on metamask, rainbow and other wallets you get `Opening metamask...` text. I think that'll probably be the best case, but ofc you can try other stuff along the way.

**Reply by TimDaub** [reply]
Well, the issue here is that I‘m fixing flaws from the Wallet ecosystem with my app.

E.g. many of my users have reported minting our NFT with their Desktop Metamask and they actually don‘t have that seed phrase on their phone. In fact, many don‘t use a phone‘s wallets.

So technically, to serve those users, when they connect on mobile they really don‘t have another choice but to „Connect with Passkeys“ where then largeBlob is then synced from iCloud. 

I either do this, or I add a second button next to „Connect Wallet,“ which would confuse everyone else though

**Reply by 0xmkh** [reply]
In that case are you able to deprecate the logic from RainbowKit into your own custom "Connect Wallet" with passkey flow ? That way you can have default wallets in RainbowKit if needed, then do your own logic with passkey and etc ? Similar to what warpcast does.

**Reply by TimDaub** [reply]
I realize now that not many know how our key management works, so here's a bit of an explanation:

- A user can use Kiwi News as any other dapp by connecting their wallet, and then for every upvote, submission, and comment, they will have to sign an EIP712 message manually.
- However, there is an optimization where we generate a new key pair in the localStorage of the users browser and we then register that key's address on a smart contract on Optimism
- Once a user has registered that key on OP mainnet, the node now recognizes signatures made from that "signer key" as valid submissions in the name of the actual custody key. And this is how we essentially make Kiwi News a "Connect Wallet" free experience. Because if the localStorage key is available, then we won't even show the "Connect" button and all signature requests go to that key
- We are purely a web app, and there is no mobile app app. And so one other issue that we ran into now is that, e.g., a user will have such a signer key in their localStorage on their Desktop PC, but on their phone, they haven't added that key to OP mainnet.
- This would lead them to have to do the same step again on their mobile phone, generate a key, store it in localStorage, and register it on OP. And that costs gas and is generally inconvenient.
- So our idea was to back up that key in iCloud with Passkeys as a largeBlob, because those keys can then be retrieved independently from the device and browser.

Hence, for these users, I think it is important to get the intent right. E.g. consider a user that only has a Desktop Metamask wallet (of which there are many). In those cases, if we could present them with an extra option after tapping "Connect" in the Wallet dialogue, this would make the most sense. They won't be able to connect their Desktop's Metamask because its not available on their mobile phone.

**Reply by 0xmkh** [reply]
Mhmm i see now. The thing is we're fully dependent on wagmi's `useAccount` hook and we look for `isConnected` boolean to close the modal immediately, this wouldn't be possible to add extra option after the "Connect" in the modal dialogue. Here is the [hook](https://wagmi.sh/react/api/hooks/useAccount#usage) for more info. There has to be a wagmi connector (wallet) which you plug into RainbowKit otherwise it won't work, and if we were to do that it would be a big migration on our end just do at that little extra step after "Connect" modal dialogue.

Are you able to create maybe your own sdk that will have another modal popup within RainbowKit's modal ? Similar to what WalletConnect and Web3Auth does ? Here is an example of web3auth's [connector](https://github.com/Web3Auth/web3auth-pnp-examples/blob/main/web-modal-sdk/wagmi-examples/rainbowkit-modal-example/src/RainbowWeb3authConnector.tsx). Maybe you can then do your own logic with localStorage, iCloud backup with passkeys, interact with smart contracts and etc...

**Reply by TimDaub** [reply]
I think I'd be capable of building that but somehow I have to understand what createConnector expects etc. Is this documented somewhere?

**Reply by 0xmkh** [reply]
Yep should be in [this](https://wagmi.sh/core/api/createConnector#createconnector) wagmi doc. Here is another example of coinbase wallet connector: https://github.com/wevm/wagmi/blob/main/packages/connectors/src/coinbaseWallet.ts

**Reply by TimDaub** [reply]
Mhh, tbh, I think it is too much more to build all of this just for trying out something. I think it'd be fine to do that if this ends up establishing itself as a rather important primitive, but we're not that far yet. As I said, I want the simplest possible way to inject code somehow into your modal. Frankly, I think the code doesn't seem to be good if it is so compartmentalized that it makes it almost impossible to do that. I just need a button with an onclick handler at the moment. Sure, we can implement it properly later, but that's not the point for reaching out ATM

**Reply by TimDaub** [reply]
Here's the demo of what I'm building. https://www.youtube.com/watch?v=_DhtZqtz_ZA

Currently I'm using the Disclaimer https://www.rainbowkit.com/docs/custom-app-info which is not optimal because e.g. the section doesn't scale to 100% width. Also, a Disclaimer element cannot have custom props defined by the application.
